# Introduction to Python

In this lesson, we will familiarize ourselves with the simplest elements of Python. We'll learn how to do basic arithmetic, manipulate variables, understand floating point and integer types, store text in string types, and define basic functions. It's a lot to take in, but we'll practice along the way!

First, some general tips for navigating this notebook.
- This notebook contains both formatted text cells (like this cell; more precisely, they are "Markdown" cells) and code cells. The code cells are preceded by the blue text `In [ ]:`
- To select a cell, click on it once. A green rectangle will appear around it.
- To execute the code in a code cell, select it (by clicking on it once) and then press Shift+Enter on your keyboard.
- Code cells can also be edited when they are selected
- Text/Markdown cells can also be edited by double-clicking on them. If you double-click on a text cell, the font will change and some strange symbols might appear. Just press Shift+Enter on your keyboard to revert to the usual display of the text

**As a general rule, you should try to guess what each code cell will do before you execute. This will enhance your understanding. Feel free to experiment with editing the contents of code cells as well! You can also insert additional cells to experiment with your own code by pressing the "+" button at the top of the page.**

## Arithmetic

We'll start by using Python as a calculator. Addition (`+`), subtraction (`-`), and multiplication (`*`) behave as you'd expect from elementary school, with the same order of operations (multiplication is evaluated before addition or subtraction). What will happen when you execute the Python code in the next cell? Select it and press Shift+Enter to find out

In [19]:
2+3*5

17

A single code cell can contain many Python expressions, each on its own line. Only the result of evaluating the final expression will be displayed in the cell's output:

In [20]:
1+1
1+2
1+3

4

If you want to display the result of multiple expressions from the same cell, you can use the `print()` command:

In [21]:
print(1-1)
1-2
print(1-3)
1-4

0
-2


-3

There are two kinds of numbers in Python: **integers** and **floating point values**. Integers behave almost exactly like the positive and negative integers you are familiar with from math classes. Floating point values are almost, but not exactly, like real numbers. Every time you see a decimal point in a number in Python, it is a floating point number rather than an integer. For example:

In [22]:
2.4*12

28.799999999999997

Notice that the expression didn't come out to exactly 28.8, as it would in real arithmetic! Arithmetic involving non-integer values is inherently imprecise in Python (and indeed in most programming languages). We'll discuss the issue at great length later on in the course.

You can also do division in Python using the forward-slash (`/`) operator. Keep in mind that the result of the `/` operator is *always* a floating point number. Notice the decimal point when you evaluate the following expression:

In [23]:
9/3

3.0

There is a separate integer division operator using two slashes: `//`. If you start with two integer values, this operator produces an integer value: specifically, it produces the greatest integer less than the result (i.e., the result of division ignoring the remainder):

In [24]:
10//3

3

(Slightly confusingly, you can also perform the `//` operator on floating point numbers. But just because you can, doesn't mean you should! If you wish, feel free to try it to deduce its function.)

What if we try to do something silly like divide by zero? Sensibly, Python throws an error:

In [26]:
1/0

ZeroDivisionError: division by zero

Getting an error (or "exception") when running Python code isn't a big deal, and it will happen frequently. Fortunately, many of the most common Python exceptions come with helpful error messages: in this case, the error message tells us that we illegally divded by zero, and points to the exact line where the error occurred.

You'll also get an error from "ungrammatical" expressions like the following:

In [27]:
10+*3

SyntaxError: invalid syntax (1200948652.py, line 1)

We mention two additional, useful operators:
- the *modulus* or *remainder* operator `%` gives the remainder after integer division
- the *exponentiation* operator `**` computes a power.

The order of operations evaluates `**` first, followed by `*`, `/`, `//`, and `%` (from left to right), followed by `+` and `-` (from left to right), more or less as in middle-school arithmetic. You can use parentheses to modify the order in which the operations are evaluated. What do you expect the following code to produce?

In [25]:
print((22+3)%11)
print(22+3%11)
print(2**3)
print(1-3**2+4//2)

3
25
8
-6


## Variables

*Variables* are used to store the results of Python expressions, so that you can reuse them in later computations. To assign a value to a variable, you use the *assignment operator* `=`, with the variable name on the left, and the value it should take on the right. For example, `pi = 3.14159` would create a variable named `pi` and give it the value `3.14159`. What do you expect the following code to do?

In [28]:
radius = 1
pi = 3.14159
area = pi*radius**2
print(area)

3.14159


You can always assign a new value to a variable. Also, within a Jupyter notebook, variables assigned in one cell are still accessible in another cell. For example, the following code runs without an error. (This is helpful for writing a program bit by bit and trying it out as you go. But it is also a **very common source of confusing bugs**: it's easy to accidentally reuse a variable with an old value you forgot you assigned.) What should the value of `circumerference` be?

In [30]:
radius = radius+1
circumference = 2*pi*radius

It's good practice to give your variables descriptive names. Vague, generic variable names like `x` or `value` make your programs hard to read. By contrast, names like `area` and `radius` convey a great deal of information about what values to expect a variable to have, making your code easier to read, modify, and debug. Variable names need to satisfy the following rules:
- use only (English) letters (upper or lowercase), numbers, and underscores (`_`) (no spaces or other special characters)
- don't start with a number

Variables that start with an underscore are allowed, but convey a special meaning and so should be avoided until you know what you're doing.

For now, we'll adopt the convention that variable names should generally be entirely lowercase, with words separated by underscores, e.g. `interest_rate` or `average_height`. Adopting a consistent convention makes code easier to read and write, because variable names are easier to predict and remember.

## Comments

Python will ignore any portion of a line following an octothorpe (`#`). This allows you to leave notes (or "comments") in your code, to make it easier to understand. It's good to get in the habit of leaving good comments in your code; its easy to forget what a particular piece of code is supposed to be doing, and if you come back to a program you wrote a week ago, or even an hour ago, it's helpful to have comments to remind you how it's supposed to work. For example, the comments, along with the descriptive variable names, (hopefully) make it clear what the following code is computing.

In [29]:
# Compute the volume of a cone with a circular base
height = 10
radius = 5 # the radius of the base
base_area = pi*radius**2 # area of a circle
volume = (1/3)*base_area*height # formula for the volume of a cone

## Strings

Python can manipulate text as easily as numbers. Text is in Python is generally represented with a **string**. Strings can be created by surrounding text with quotes (either double `"` or single `'`, but use the same kind of quotes on each side).

In [31]:
course_name = "Applied Linear Algebra Practicum"
course_room = 'Goldsmith 226'

Strings have their own rules for arithmetic. If you combine two strings with the `+` operator, they will be *concatenated*:

In [32]:
print(course_name + " is held in " + course_room)

Applied Linear Algebra Practicum is held in Goldsmith 226


You can also multiply strings by integers to repeat them a number of times:

In [33]:
"Knock"*2

'KnockKnock'

You cannot, however, add strings and numbers together. This is a good thing, because the result would be ambiguous:

In [38]:
"100"+1 # Should this be the string "1001" or the integer 101? Python gives an error

TypeError: can only concatenate str (not "int") to str

To combine strings and numbers (using concatenation or numerical addition) you need to first convert either the number to a string (for concatenation) or the string to a number (for numerical addition). Such conversion can be done using the `str`, `int`, and `float` functions:

In [39]:
print("100"+str(1)) # Convert the integer to a string, and concatenate
print(int("100")+1) # Convert the string to an integer, and add
print(float("100")+1) # Convert the string to a floating point number, and add

1001
101
101.0


(You can also use the `int` and `float` functions to convert between integer and floating point numerical types:)

In [40]:
print(int(pi)) # Python truncates to the greatest integer less than pi

3


## Functions

A variable is value that can be reused over and over again; a *function* is an entire block of code that can be reused over and over again. Let's see an example first, before we carefully go over the syntax and terminology surrounding functions.

In [ ]:
# We define a function called `greet` with a single parameter, `name`
def greet(name):
    # This function prints a greeting, personalized with the name parameter
    print("Hello there, "+name+"!")
    
# We now call the function twice, with different arguments
greet("Alice")
greet("Bob")

In the code above, we first *defined* a function, to be used later. The first line of the function definition has four pieces:
1. The special keyword `def` that is used whenever you define a function
2. The name of this particular function (in this case `greet`). This is like a variable name; it's what we'll use to refer to the function later on. The same rules apply to function names as to variable names (e.g., only use letters, numbers, and underscores).
3. A (comma-separated) list of function *parameters*, surrounded by parentheses. This list could be empty if the function doesn't take parameters. In this case, there is a single parameter (`name`) (so we didn't need any commas). We'll see an example later with more than one parameter. The parameters are like variables that can be used within the function.
4. A colon (`:`) at the end.

After the first line of the function definition, we have one or more lines of code that will be executed every time the function is called. These lines of code must all be indented using the same amount of whitespace; the usual convention is to use four spaces for the indentation. Within the Jupyter notebook, you can create those four spaces by pressing the Tab key. The function `greet` above has two lines of code in its definition, a comment and a `print` statement.

Finally, after defining the function, we can *call* it to execute its code. Every time we call a function, we pass *arguments* to assign values for each of its parameters. These arguments appear in a comma-separated list surrounded by parentheses, in the same order as the parameter definition. In the above code, we first call the `greet` function with the `name` parameter equal to the value `"Alice"`, and then call it again with the `name` parameter assigned the value `"Bob"`.

Most functions will compute some value or values that you will want to use in other parts of your program. These values need to be *returned* once they are computed. To return a value, use the special `return` keyword, followed by the value you want returned. You can think of the parameters of a function as its input, and its return value as its output. Note that a function will stop executing once it reaches a return statement.

Before giving another example of a function, we first explain another construction that works very much like a comment in Python: a *docstring*. The convention in Python is to document each function you define by writing a string surrounded by three quotes (`"""like this"""`) on the first line of the function (right after the line with `def`). Usually the docstring will briefly describe what the function does, explain any parameters the function takes, and describe any values that are returned.

In the following example, we have a function that has two parameters and returns a value. We have used a (rather extravagent) docstring to document it (the first six lines of the function)

In [41]:
def cone_volume(base_area, height):
    """Compute the volume of a cone with a given base area and height.
    
        base_area - the area of the base of the cone
        height - the height of the cone
        
        Returns the volume of the cone"""
    return (1/3)*base_area*height

volume = cone_volume(5, 2) # Call the function cone_volume, and assign its return value to the variable `volume`
print(volume)

3.333333333333333


## Terminology

We've covered a lot of ground here! In particular, we've introduced quite a bit of jargon. Let's now summarize all the terms used so far, and introduce some additional relevant terminology:


- A *comment* is text following a `#` symbol that is not executed by Python, used to describe and explain the surrounding code
```python
# This is a comment.
```
- A *value* is a single unit of data in your code. For example, `12` is an integer value, and `"hello"` is a string value
- A *statement* is a grammatically sensible piece of code
```python
x = 12 # A statement.
print(x) # Another statement.
x+3 # Yet another statement.
```
- An *expression* is a statement that, when run, evaluates to a single value
```python
4*(12//3) # An expression that evaluates to the integer 16
str(1)+"+"+str(2)+"="+str(1+2) # An expression that evaluates to the string "1+2=3"
```
- A *variable* is container that associates a single value with a name so that it can be reused
```python
age = 19
# age is a variable. Its value is an integer
```
- A *string* is a value representing text
- An `int` is a type of numerical value representing positive or negative integers. You can expect Python to perform arithmetic with `int` values exactly
- A `float` is a type of numerical value representing real numbers. You should expect any arithmetic you perform with `float` values to give inexact (but reasonably accurate) results
- A *function* is an organizational unit of reusable code. Functions are *defined* using a statement starting with the `def` keyword, followed by one or more indented lines of code
- - Functions can have zero or more *parameters*. A parameter is a variable whose value is assigned when the function is called.
- - Values can be *passed* to a function as input when the function is called. These values are called *arguments*; each argument is assigned to a parameter variable
- - A function can *return* a value as output. When the function is called, it evaluates to its return value.
```python
def function_name(parameter1, parameter2):
        """The docstring goes here and explains the function."""
        print("The first line of this function prints something.")
        x = parameter1 + parameter2 # The second line of the function
        return x*2 # A return statement. The value x*2 is returned by this function
function_name(1, 2)
# The function is called with arguments 1 and 2, which are assigned to parameter1 and parameter2.
# The above function call evaluates to 6
```